runtime/queries/metricsview_comparision_toplist.go

```diff
@@ -0,0 +1,280 @@
+			DimensionValue: dv,
+			MeasureValues:  measureValues,
+		})
+	}
+
+	q.Result = &runtimev1.MetricsViewComparisonToplistResponse{
+		Data: data,
+	}
+
+	return nil
+}
+
+func timeRangeClause(start, end *timestamppb.Timestamp, td string, args *[]any) string {
+	var clause string
+	clause += fmt.Sprintf(" AND %s >= ?", td)
+	*args = append(*args, start.AsTime())
+
+	clause += fmt.Sprintf(" AND %s < ?", td)
+	*args = append(*args, end.AsTime())
+
+	return clause
+}
+
+func (q *MetricsViewComparisonToplist) buildMetricsTopListSQL(mv *runtimev1.MetricsView, dialect drivers.Dialect) (string, []any, error) {
+	dimName := safeName(q.DimensionName)
+	selectCols := []string{dimName}
+	finalSelectCols := []string{}
+	measureMap := make(map[string]int)
+	for i, n := range q.MeasureNames {
+		measureMap[n] = i
+		found := false
+		for _, m := range mv.Measures {
+			if m.Name == n {
+				expr := fmt.Sprintf(`%s as "%s"`, m.Expression, m.Name)
+				selectCols = append(selectCols, expr)
+				finalSelectCols = append(
+					finalSelectCols,
+					fmt.Sprintf(
+						"base.%[1]s, comparison.%[1]s, comparison.%[1]s - base.%[1]s, (comparison.%[1]s - base.%[1]s)/base.%[1]s::DOUBLE",
+						m.Name,
+					),
+				)
+				found = true
+				break
+			}
+		}
+		if !found {
+			return "", nil, fmt.Errorf("measure does not exist: '%s'", n)
+		}
+	}
+	subSelectClause := strings.Join(selectCols, ", ")
+	finalSelectClause := strings.Join(finalSelectCols, ", ")
+
+	baseWhereClause := "1=1"
+	comparisonWhereClause := "1=1"
+
+	args := []any{}
+	if mv.TimeDimension == "" {
+		return "", nil, fmt.Errorf("Metrics view '%s' doesn't have time dimension", mv.Name)
+	}
+
+	td := safeName(mv.TimeDimension)
+
+	baseWhereClause += timeRangeClause(q.BaseTimeStart, q.BaseTimeEnd, td, &args)
+	if q.Filter != nil {
+		clause, clauseArgs, err := buildFilterClauseForMetricsViewFilter(q.Filter, dialect)
+		if err != nil {
+			return "", nil, err
+		}
+		baseWhereClause += " " + clause
+
+		args = append(args, clauseArgs...)
+	}
+
+	comparisonWhereClause += timeRangeClause(q.ComparisonTimeStart, q.ComparisonTimeEnd, td, &args)
+	if q.Filter != nil {
+		clause, clauseArgs, err := buildFilterClauseForMetricsViewFilter(q.Filter, dialect)
+		if err != nil {
+			return "", nil, err
+		}
+		comparisonWhereClause += " " + clause
+
+		args = append(args, clauseArgs...)
+	}
+
+	orderClause := "true"
+	for _, s := range q.Sort {
+		i := measureMap[s.MeasureName]
+		orderClause += ", "
+		var pos int
+		switch s.Type {
+		case runtimev1.ComparisonSortType_COMPARISON_SORT_TYPE_BASE_VALUE:
+			pos = 2 + i*4
+		case runtimev1.ComparisonSortType_COMPARISON_SORT_TYPE_COMPARISON_VALUE:
+			pos = 3 + i*4
+		case runtimev1.ComparisonSortType_COMPARISON_SORT_TYPE_DELTA:
+			pos = 4 + i*4
+		default:
+			return "", nil, fmt.Errorf("undefined sort type for measure %s", s.MeasureName)
+		}
+		orderClause += fmt.Sprint(pos)
+		if !s.Ascending {
+			orderClause += " DESC"
+		}
+		orderClause += " NULLS LAST"
```

Comment:
Need a conditional to only add this for DuckDB, it's not supported by Druid (see old toplist query)